pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go

```diff
@@ -137,6 +162,109 @@ func (grb *globalRoleBindingLifecycle) deleteAdminBinding(obj *v3.GlobalRoleBind
 	return nil
 }
 
+// reconcileClusterPermissions grants permissions for the binding in all downstream (non-local) clusters. Will also
+// remove invalid bindings (bindings not for active RoleTemplates or for invalid subjects).
+func (grb *globalRoleBindingLifecycle) reconcileClusterPermissions(globalRoleBinding *v3.GlobalRoleBinding) error {
+	globalRole, err := grb.grLister.Get("", globalRoleBinding.GlobalRoleName)
+	if err != nil {
+		return fmt.Errorf("unable to get globalRole %s: %w", globalRoleBinding.Name, err)
+	}
+	// no per-cluster permissions if the associated globalRole doesn't inherit any clusterRoles
+	if len(globalRole.InheritedClusterRoles) == 0 {
+		return nil
+	}
```

Comment:
If a previous generation had "corrupt roles" and the new generation has no inherited cluster roles, the short circuit here may remove the ability for the controller to self-heal (a.k.a. purge the corrupt roles).

Reply:
You are correct, I've made an update to attempt to purge roles first. I've also augmented this to purge roles that are attempting to target the local cluster (in case a user manually makes one of those).